### ISSUE TITLE
Dropped response bodies when using Envoy as an HTTP CONNECT proxy

### ISSUE THREAD

**Issue by nicktrav** [envoyproxy/envoy#12996]
*Description*:

I'm using Envoy as an HTTP CONNECT proxy, and observing dropped response bodies. This happens non-deterministically - sometimes the client gets the response payload, sometimes it doesn't.

Specifically, Envoy seems to be associating the response payload from the upstream server with the HTTP response it initially sends back to the client as part of the CONNECT protocol.

[RFC 7231, section 4.3.6](https://tools.ietf.org/html/rfc7231#section-4.3.6) has this to say about what this should happen:

> Any 2xx (Successful) response indicates that the sender (and all
> inbound proxies) will switch to tunnel mode immediately after the
> blank line that concludes the successful response's header section;
> data received after that blank line is from the server identified by
> the request-target.

The HTTP response should not contain anything after the blank line following the headers. I'm observing cases where it does.

*Repro steps*:

Using the following config:

```yaml
admin:
  access_log_path: /dev/stdout
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 9901

static_resources:
  listeners:
  - name: listener_http
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 8080
    access_log:
      name: envoy.access_loggers.file
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
        path: /dev/stdout
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains:
              - "*"
              routes:
              - match:
                  connect_matcher:
                    {}
                route:
                  cluster: greeter
                  upgrade_configs:
                  - upgrade_type: CONNECT
                    connect_config:
                      {}
          http_filters:
          - name: envoy.filters.http.router

  clusters:
  - name: greeter
    connect_timeout: 0.25s
    type: STRICT_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: greeter
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: localhost
                port_value: 1234
```

I'm running the proxy in a container bound to the host network (have also tried outside a container, and I see the same thing):

```
$ docker run --rm -it \
  --network host \
  -v $(pwd)/envoy.yaml:/etc/envoy/config.yaml \
  envoyproxy/envoy-dev:latest -c /etc/envoy/config.yaml
```

Set up a simple server that prints a message on session establishment:

```shell
$ while true; do echo 'hello, world' | nc -l 1234; done
```

Simulate an HTTP CONNECT from a client:

```shell
$ printf 'CONNECT localhost:1234 HTTP/1.1\nHost: localhost:1234\n\n' | nc -w1 localhost 8080
```

Observe the response traffic from Envoy in tcpdump  / wireshark.

Case a) (unexpected behavior) - proxy returns a 200 to the client _along with the response payload from the upstream_.

![Screenshot from 2020-09-06 15-29-57](https://user-images.githubusercontent.com/804457/92336485-134eb380-f056-11ea-8c42-b5b2de79fcae.png)

Frame 402:

![Screenshot from 2020-09-06 15-30-23](https://user-images.githubusercontent.com/804457/92336487-13e74a00-f056-11ea-8953-95a452fa7df1.png)

Case b) (expected behavior) - proxy returns a 200 to the client without a body, followed by a packet containing the response payload from the upstream (i.e. payload is not part of the "HTTP body").

![Screenshot from 2020-09-06 15-30-43](https://user-images.githubusercontent.com/804457/92336488-18136780-f056-11ea-8a46-8d88a78dfb45.png)

Frame 904:

![Screenshot from 2020-09-06 15-30-58](https://user-images.githubusercontent.com/804457/92336489-18abfe00-f056-11ea-945f-d9276504bace.png)

Frame 906:

![Screenshot from 2020-09-06 15-31-07](https://user-images.githubusercontent.com/804457/92336490-19449480-f056-11ea-91f2-7220dd5c4860.png)

Repeatedly issuing requests causes instances of both - there is no discernible pattern that I could tell.

Version information:

```shell
$ envoy --version
envoy  version: 2dfaf6eb19df6e88fff9563470565ec37c61b23d/1.16.0-dev/Clean/RELEASE/BoringSSL

$ uname -a
Linux nickt 5.4.0-45-generic #49-Ubuntu SMP Wed Aug 26 13:38:52 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by mattklein123** [comment]
cc @alyssawilk. I'm guessing any issue with timing depending on when the data from upstream arrives. Thanks for the detailed report!

**Comment by alyssawilk** [comment]
Sorry, reading through this I want to make sure I understand the problem
Is the problem that you want
[envoy headers] 
[helloworld payload] 
in separate packets and sometimes you get
sometimes you get
[envoy headers][helloworld payload]
in one packet?

If so, I don't think that's illegal.  I think anything after the \r\n\r\n headers is payload content regardless of it lands in the same packet or not.

**Comment by nicktrav** [comment]
> Is the problem that you want

Hey @alyssawilk - yeah, that's right.

> If so, I don't think that's illegal. I think anything after the \r\n\r\n headers is payload content regardless of it lands in the same packet or not.

Agreed on the semantics there.

The reproducer I mentioned was more to show the TCP framing differences (two packets in the response vs. one, as you mention). My true setup is a little more complex:

```
client -> Go client -> Envoy (CONNECT passthrough) -> Envoy (CONNECT termination) -> upstream
```

Based on the framing, I realized the issue is in the Go `net` package, which doesn't really support CONNECT all that well, and is gobbling up the response body in the case of a single frame. Checking for a body and writing it back to the network fixed my issue.

Closing this, as Envoy isn't really the issue here :smile:

Thanks!

---

For anyone that comes across this in the future by way of Go / HTTP, here are some relevant links / advice:

> do *http.Request.Write it to a self-dialed net.Conn, rather than using *http.Transport.

https://github.com/golang/go/issues/22554#issuecomment-341789905

> People today typically do something like this:
> https://github.com/golang/build/blob/e12c9d226b16d4d335b515404895f626b6beee14/cmd/buildlet/reverse.go#L197

https://github.com/golang/go/issues/32273#issuecomment-496722323
